pkg/reconciler/namespacecontroller/state.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// State stores the state of the Namespace Controller.
+type State struct {
+	// eventMux is a RMMutex to protect read/write to 'isSyncingPending'.
+	// It is RWMutex because the cache is read more often than writen.
+	eventMux sync.RWMutex
+
+	// isSyncPending indicates whether the namespace controller fires up an event that requires a new sync.
+	isSyncPending bool
+
+	// selectorMux is the mutex to protect read/write to 'nsSelectors' and 'selectedNamespaces'.
+	selectorMux sync.Mutex
+
+	// nsSelectors is a map of NamespaceSelector names to the LabelSelectors.
+	nsSelectors map[string]labels.Selector
+
+	// selectedNamespaces records the dynamic/on-cluster Namespaces selected by NamespaceSelectors
+	selectedNamespaces map[string][]labels.Selector
+}
+
+// NewState instantiates the namespace controller state.
+func NewState() *State {
+	return &State{
+		nsSelectors:        make(map[string]labels.Selector),
+		selectedNamespaces: make(map[string][]labels.Selector),
+	}
+}
+
+// setSyncPending indicates an event is detected to trigger a new sync.
+func (s *State) setSyncPending() {
+	s.eventMux.Lock()
+	defer s.eventMux.Unlock()
+	s.isSyncPending = true
+}
+
+// IsSyncPending returns whether the state triggers a sync.
+func (s *State) IsSyncPending() bool {
+	s.eventMux.Lock()
+	defer s.eventMux.Unlock()
+	isSyncingPending := s.isSyncPending
+	if s.isSyncPending {
+		// Reset the Sync pending state after a sync regardless if succeeds or not.
+		// If it fails, a retry should be triggered.
+		s.isSyncPending = false
```

Comment:
nit: the function signature `IsSyncPending` gives the impression this is a simple getter style method, when this in reality may mutate state.

Suggest a change in the name to include a verb, for example `StartResync`

Reply:
Renamed to `ScheduleSync`. 

`StartResync` seems to belong to a method of the reconciler thread, not the state.

Reply:
Sounds good, thank you